[17] account_reconcile_oca : Forward port multi currency management from 16

florian-dacosta opened the pull request:
@etobella @pedrobaeza 

Pre-commit is failing because of a part coming from https://github.com/OCA/account-reconcile/pull/694/commits/62d508bb2077e22683883713fd1dadd1430f6d73#diff-4032c7e70261a4eb503cedc278a75ad0b0d8e20843b621b4e9bae34a2426f153
I am not comfortable at all with this part, could you take a look @etobella ?

Forward port of https://github.com/OCA/account-reconcile/pull/694
Include also : 
- https://github.com/OCA/account-reconcile/pull/787
- https://github.com/OCA/account-reconcile/pull/792
- https://github.com/OCA/account-reconcile/pull/784

Multi-currency is still quite buggy, but it makes main use cases ok at least.

etobella committed:
[IMP] account_reconcile_oca: Improve multicurrency management.

We will use currency of the line in order to get the suspense and max line value

florian-dacosta committed:
[FIX] account_reconcile_oca : Fix multi currency management
Fix the case of payment with a foreign currency set on bank statement line

improve tests around multi-currency

florian-dacosta committed:
[FIX] account_reconcile_oca : foreign currency reconcile with late currency rate
It is possible that the statement line in foreign currency is created before the rate of the day is updated in Odoo. In this case we need to take the real rate of the statement line to comput the exchange rate

florian-dacosta committed:
[FIX] account_reconcile_oca : max_amount rounding error leading to unwanted currency conversion

florian-dacosta committed:
[FIX] account_reconcile_oca : exchange rate gain/loss currency and amount
If currency amount is not 0, the suspense line will have wrong amount

etobella committed:
[FIX] account_reconcile_oca: Update currency_amount

victoralmau committed:
[FIX] account_reconcile_oca: Synchronize manual_partner_id with partner_id without onchange or subsequent changes with the _synchronize_to_moves() method.

We do not define partner_id with the value of manual_partner_id to prevent _synchronize_to_moves()
from making changes to the account.move.line leaving unintended values and/or data.

Re-define the value of reconcile_data_info if the _synchronize_to_moves() method has changed
anything on the lines.

Related to https://github.com/OCA/account-reconcile/issues/779

TT52634

florian-dacosta committed:
[FIX] Synchronise partner from statement line to accounting entries
We need to look at the reconcile_data to find the partner as the manual_partner_id can contain the information of an auxiliary line

florian-dacosta committed:
[FIX] fix _check_line_changed partner check

florian-dacosta committed:
[FIX] Wrong amount currency compute in case of change in liquidity line

sergiobstoj committed:
[FIX] account_reconcile_oca: Fix currency and amount when using reconciliation models

victoralmau reviewed (changes requested):
Code review OK, please, fix pre-commit errors.,

florian-dacosta commented:
Yes, I am waiting @etobella On this one!

etobella commented on account_reconcile_oca/models/account_bank_statement_line.py:
```diff
@@ -570,16 +635,87 @@ def _default_reconcile_data(self, from_unreconcile=False):
+                    partial = partial_lines.filtered(
+                        lambda r: r.debit_move_id == reconciled_line
+                        or r.credit_move_id == reconciled_line
+                    )
```
```suggestion
                    partial = partial_lines.filtered(
                        lambda r, line=reconciled_line: r.debit_move_id == line
                        or r.credit_move_id == line
                    )
```

Doing something like this will fix the pre-commit

victoralmau commented:
Can you also add the changes from https://github.com/OCA/account-reconcile/pull/792?

florian-dacosta commented:
Pre-commit is fixed and and https://github.com/OCA/account-reconcile/pull/792 added.

florian-dacosta committed:
[FIX] partner update from reconciliation widget
We cant to avoid to change amounts on accounting entries during the reconciliation process. One of the goal is to avoid a following case :
The statement line is created in a foreign currency journal, later, the rate is updated in Odoo.
During reconciliation process, the partner is set on liquidity line, the accounting entries are synchronized and the balance changes.

pedrobaeza commented:
Please squash some commits together like the linter fix with the code that originates the linter the error for merging this.

florian-dacosta committed:
[FIX] Replace removed compute method by _convert

florian-dacosta commented:
> Please squash some commits together like the linter fix with the code that originates the linter the error for merging this.

I have squashed this one.
I have also added https://github.com/OCA/account-reconcile/pull/784
and fix tests with last commit (unrelated to any other commit, it is coming from the migration of the module I think)

pedrobaeza reviewed (approved):
/ocabot merge patch
